### PR TITLE
feat: Implement parcel decryption

### LINKED ISSUE #10: Implement decryption of incoming parcels
This Part 2 of the `POST /` endpoint in the PoHTTP server.

Example: https://github.com/relaycorp/relaynet-pong/blob/6bc207a06458a3f66cf324cc99d824df29781cd7/src/app/background_queue/PingProcessor.ts#L71-L80

This is the part that will create or update a channel between the Internet endpoint and a peer.

See also:

- Part 1: #9.
- Part 3a: #15.
- Part 3b: #11.

### PR DESCRIPTION
**Pull request by levansuper**
Fixes #10.

### CODE REVIEW

File: src/server/routes/parcel.routes.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+          .send({ message: 'Parcel is well-formed but invalid' });
+      }
+
+      let decryptionResult;
+      try {
+        decryptionResult = await parcel.unwrapPayload(activeEndpoint.keyStores.privateKeyStore);
+      } catch (error) {
+        parcelAwareLogger.info({ err: error }, 'Invalid service message');
+        return reply
+          .code(HTTP_STATUS_CODES.BAD_REQUEST)
```
Comment (gnarea) [inline src/server/routes/parcel.routes.ts:62]: Note that in Pong, we would've returned a `HTTP_STATUS_CODES.ACCEPTED` before even attempting to decrypt the payload. We should do that here too: The HTTP client isn't responsible for the E2E encrypted payload -- that's the responsibility of the endpoint that created the parcel.

```suggestion
          .code(HTTP_STATUS_CODES.ACCEPTED)
```

File: src/server/routes/parcel.routes.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        return reply
+          .code(HTTP_STATUS_CODES.FORBIDDEN)
+          .send({ message: 'Parcel is well-formed but invalid' });
+      }
+
+      let decryptionResult;
+      try {
+        decryptionResult = await parcel.unwrapPayload(activeEndpoint.keyStores.privateKeyStore);
+      } catch (error) {
+        parcelAwareLogger.info({ err: error }, 'Invalid service message');
```
Comment (gnarea) [inline src/server/routes/parcel.routes.ts:60]: Nitpick

```suggestion
      } catch (err) {
        parcelAwareLogger.info({ err }, 'Invalid service message');
```

File: src/server/routes/parcel.routes.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+          .code(HTTP_STATUS_CODES.BAD_REQUEST)
+          .send({ message: 'Invalid service message' });
+      }
+
+      // This log is needed not to throw decryptionResult is unused error.
+      // Will be removed in the next PR
+      parcelAwareLogger.info(
+        { test: decryptionResult.senderSessionKey.keyId },
+        'Invalid service message',
+      );
```
Comment (gnarea) [inline src/server/routes/parcel.routes.ts:71]: Feel free to keep it here, but please add a note to that issue so we won't miss it. As a reviewer, it'd be quite easy for me to miss it.

Alternatively, remove these lines and the variable `decryptionResult` above -- you don't need that to run the try/catch.

**Review by gnarea — CHANGES_REQUESTED** [review]
This PR is bringing a lot of the code that was already implemented in #27, so I couldn't do a proper review: I only focused on the decryption logic in `src/server/routes/parcel.routes.ts`.

Please fix the bad merge and I'll do a more thorough review.

File: src/server/routes/parcel.routes.ts (hunk 3 above)
Comment (levansuper) [reply src/server/routes/parcel.routes.ts:70]: added comment here
[Store incoming PDAs#15](https://github.com/relaycorp/awala-endpoint-internet/issues/15)

File: src/server/routes/parcel.routes.spec.ts [hunk 4]
```diff
@@ -115,4 +122,26 @@ describe('parcel route', () => {
     );
     expect(logs).toContainEqual(partialPinoLog('info', 'Refusing invalid parcel'));
   });
+
+  test('Invalid service message should be accepted', async () => {
```
Comment (gnarea) [inline src/server/routes/parcel.routes.spec.ts:126]: "accepted" is ambiguous here. I'd say "ignored".

File: src/server/routes/parcel.routes.spec.ts [hunk 5]
```diff
@@ -115,4 +122,26 @@ describe('parcel route', () => {
...
+      new Date(),
+    );
+
+    const response = await server.inject({
+      ...validRequestOptions,
+      payload: parcelSerialized,
+    });
+
+    expect(response).toHaveProperty('statusCode', HTTP_STATUS_CODES.ACCEPTED);
+    expect(logs).toContainEqual(partialPinoLog('info', 'Invalid service message'));
```
Comment (gnarea) [inline src/server/routes/parcel.routes.spec.ts:145]:
```suggestion
    expect(logs).toContainEqual(partialPinoLog('info', 'Ignoring invalid service message'));
```

File: src/testUtils/awala/parcel.ts [hunk 6]
```diff
@@ -1,19 +1,50 @@
...
+function serializeMessage(pdaPath: CertificationPath, endpointInternetAddress: string): Buffer {
+  const message = {
+    id: randomUUID(),
+    // eslint-disable-next-line @typescript-eslint/naming-convention, camelcase
+    pda_path: Buffer.from(pdaPath.serialize()).toString('base64'),
+    // eslint-disable-next-line @typescript-eslint/naming-convention, camelcase
+    endpoint_internet_address: endpointInternetAddress,
+  };
+  return Buffer.from(JSON.stringify(message));
+}
```
Comment (gnarea) [inline src/testUtils/awala/parcel.ts:35]: This function adds too much complexity and is unnecessary in this repo. We're not processing pings here.

We should just create a stub `SERVICE_MESSAGE_CONTENT` (`Buffer`) and use it everywhere.

File: src/testUtils/awala/parcel.ts [hunk 7]
```diff
@@ -29,31 +60,48 @@ async function generateStubNodeCertificate(
   });
 }
 
-async function generateParcelPayload(recipientIdCertificate: Certificate): Promise<Buffer> {
+async function generateSessionlessParcelPayload(
+  recipientIdCertificate: Certificate,
+): Promise<Buffer> {
   const serviceMessageEncrypted = await SessionlessEnvelopedData.encrypt(
```
Comment (gnarea) [inline src/testUtils/awala/parcel.ts:66]: I missed this in a previews review.

There such be no use of of the `SessionlessEnvelopedData` anywhere in this repo. As you may remember from an exception you shared with me, that class is in the process of being removed across the board.

We should be using `SessionEnvelopedData` exclusively.

**Review by gnarea — CHANGES_REQUESTED** [review]
Please be careful not to port anything related to ping or pong service messages in the future. That is the only business logic from the Pong repo that we do not want here. I added notes where that was done.

File: src/testUtils/awala/parcel.ts (hunk 6 above)
Comment (levansuper) [reply src/testUtils/awala/parcel.ts]: The thing is I didn't understand the structure of the message. After working on the PDA pr I got good sense of it. Reworked this part

**Review by gnarea — CHANGES_REQUESTED** [review]
Nearly there. Just a minor suggestion that I'll implement myself in a moment, so I can merge this PR and then start #11.

**Review by gnarea — APPROVED** [review]

**Pull request merged**